lib/spack/llnl/util/lang.py

```diff
@@ -368,10 +368,8 @@ def dedupe(sequence):
         stable de-duplication of the sequence
     """
     seen = set()
-    for x in sequence:
-        if x not in seen:
-            yield x
-            seen.add(x)
+    seen_add = seen.add
+    return [x for x in sequence if not (x in seen or seen_add(x))]
```

Comment:
At a first glance the order before and after this PR should be the same, as we are iterating through `sequence` in the same order. The only difference I see here is that the function is no more a generator, but it returns directly a fully-fledged list.

The conditional in my opinion reads better in the current state, as in the list comprehension one is obliged to test:
```python
if not (x in seen or seen_add(x))
```
and rely on the fact that a function without a return value (`set.add`) returns `None`. This is done to be able to add a function call with side-effects within the list comprehension.

Reply:
I agree that the new function takes advantage of a not-usually-Python-esque idiom, but if @gartung has verified that it is the current `dedupe()` compromises the ordering rather than anything upstream, then we need something that does not. I'm not sure whether one could restore the generator expression character simply by replacing `[]` with `()`, but it is the case that every existing use of `dedupe()` calls it in conjunction with `list()`, thereby negating the generator character instantly.

Reply:
OK. So the problem is higher up in the stack because the list order is not preserved before dedupe() is called.

Reply:
@chissg I am not advocating against turning `dedupe` into a regular function. Just saying that I don't see where in these few lines the order changes (as claimed in the title of the PR). Asking because if the ordering is the same, the bug might be lurking somewhere else - maybe a missing wrapping `list` around a `dedupe` statement? - and it would be good to know that.

Also, if we go with this, I think it we should unwrap all the uses of `dedupe` from the enclosing `list()` constructor as they are not necessary anymore.

Reply:
I am now unable to reproduce the problem with a clean checkout of the head of develop. I suspected it might have to do with the caching of patches and switching between the fnal-develop branch and the spack develop branch.

Reply:
From comparing the patch cache I see that the order is not preserved.